### ISSUE TITLE
[RFC] 4.0.0 Release

### ISSUE THREAD

**Issue by jameslamb** [microsoft/LightGBM#5153]
## Summary

@StrikerRUS @guolinke @shiyu1994 @jmoralez @btrotta @Laurae2 could you please try to list out what you all feel is required before a v4.0.0 release of LightGBM is prepared?

Please edit this description and add issues that you feel should be fixed prior to a v4.0.0 release. I've proposed an initial list below.

**Python package**

* [ ] #4220
* [ ] #4771
* [x] #4788
* [x] #4793
* [x] #5313
* [x] #4518
* [x] #3556
* [x] #5354
* [x] #5027
* [x] #5105
* [x] #4596
* [x] #5052

**R package**

* [ ] #4323
* [ ] #4904
* [ ] #4970
* [ ] #5102
* [ ] #5135
* [x] #4881
* [x] #5036
* [x] #4966
* [x] #4982
* [x] #4813
* [x] #5108
* [x] #5116
* [x] #5122
* [x] #4670

**CUDA**

* [ ] #4781
* [ ] #5062
* [ ] #5076
* [ ] #5086
* [ ] #5163
* [x] #4785

**GPU (non-CUDA)**

* [ ] #3339
* [ ] #3475
* [ ] #3645
* [ ] #3870
* [ ] #4082
* [ ] #4618
* [ ] #4612
* [ ] #4820
* [x] #4497

**Quantized Training**
* [ ] #5606 

**Other**

* [ ] #3234
* [ ] #3679
* [ ] #4156
* [ ] #4286
* [ ] #4331
* [ ] #4405
* [ ] #4414
* [ ] #4739
* [ ] #4780
* [ ] #4946
* [ ] #5093
* [ ] #5124
* [ ] #5091 
* [x] #4827
* [x] #5391
* [x] #5344
* [x] #4649

## Motivation

The most recent release of LightGBM, v3.3.1, was about 6 months ago ([October 27, 2021](https://github.com/microsoft/LightGBM/releases/tag/v3.3.1).

*There was a v3.3.2  release on January 7, 2022, but it just contained a single small patch to satisfy CRAN*

We have been talking for even longer than that about putting out a 4.0.0 release of LightGBM. Many many fixes, features, and breaking changes have been merged since v3.3.1, and the state of this project on `master` is now significantly different from what users will get running `pip install lightgbm`.

I'd like to try to list out everything that we feel needs to be done before the 4.0.0 release, to focus our efforts and hopefully get that release out to users soon. I think an issue like this was successful with the v3.3.0 release (#4310).

## Other Items for Release Checklist

* [x] enable building `stable` tag on readthedocs

**Comment by jameslamb** [comment]
For now, I've locked the conversation on this issue so that only collaborators in the repo are able to comment, to keep the conversation focused on what maintainers are comfortable committing to for v4.0.0.

Users and other outside interested parties can open other issues referencing this one with questions or concerns.

**Comment by StrikerRUS** [comment]
Great plan!
But sorry, I don't believe we'll able to release `v.4.0.0` in the near future with the current project activity level.

I strongly believe we should focus on bug fixes and reviewing PRs from an outside contributors to keep users' loyalty.

**Comment by jameslamb** [comment]
> I don't believe we'll able to release v4.0.0 in the near future with the current project activity level

I also feel that it's probably far away given the current project activity level, but I wanted to at least *try* to push v4.0.0 forward. I hope maybe the conversation on this issue will encourage Microsoft and other maintainers here to devote more time and resources to making it happen.

I expect that many (maybe even most) LightGBM users won't rely on specific commits, either via `git clone` or nightly builds, that aren't tagged as releases or published to package repositories like CRAN, PyPI, and conda-forge (which just re-bundles the PyPI package). For them, improvements to this project aren't "real" until they make it into a release.

And I can say that for me personally, it's difficult to stay motivated to spend time on challenging bug fixes, pull request reviews, and user questions in issues if it seems like that work might not ever be released 😞

**Comment by guolinke** [comment]
For the GPU(non-CUDA) bugs, I think it is hard to fix them, as we cannot reach its original developer.
I think we can focus on the cuda_exp version, and deprecate the old GPU codes.

**Comment by guolinke** [comment]
How about we make two lists? one is for necessary changes, another one is optional. 
we can focus on the bug fixes, breaking but necessary new features/changes.
We can also mark the assignees in the list (you can also include me), and I will try my best to finish them.

**Comment by jameslamb** [comment]
> How about we make two lists?

My goal with this issue was to define the list of "what is required for v4.0.0", and I think that's enough. The implication of that would be that anything not listed here is optional.

**Comment by guolinke** [comment]
@jameslamb haha, there are 52 items, so I think maybe it is too large, and some "hard" items may block us to release. So I propose to have a "core" list that must be done for the v4.0.0.

**Comment by StrikerRUS** [comment]
@guolinke 

> I think we can focus on the cuda_exp version, and deprecate the old GPU codes.

I'm very disappointed in this decision 🙁 

[As I said earlier](https://github.com/microsoft/LightGBM/pull/4630#discussion_r718064513), OpenCL-based version that is able to run on AMD and Intel GPUs is a competitive advantage of LightGBM.

> For the GPU(non-CUDA) bugs, I think it is hard to fix them, as we cannot reach its original developer.

Have you tried to reach him via email?

**Comment by guolinke** [comment]
@StrikerRUS for the AMD GPU, we can use rocm, and there are several tools that can covert CUDA code to rocm code. 
As for Intel GPU, I am not sure, are there any powerful GPUs that widely used?

**Comment by shiyu1994** [comment]
@jameslamb Thanks for opening this! I'm glad and exciting to work towards 4.0.0. I'll commit more time from now on to guarantee the progress.

> I think we can focus on the cuda_exp version, and deprecate the old GPU codes.

We will make cuda_exp version supporting as many platforms as now by LightGBM. However, before cuda_exp is stable enough, I think we can still maintain the current CUDA and GPU versions, including bug fixing.

**Comment by StrikerRUS** [comment]
> for the AMD GPU, we can use rocm, and there are several tools that can covert CUDA code to rocm code.

Oh, interesting idea, I've never heard about such tools! I was sure that CUDA can be run exclusively on NVIDIA cards. However, do you think code transpilation will be convenient for LightGBM users? Honestly, I don't believe that ordinary users will do such things. In contrast, OpenCL version can be used just out of the box [on Windows](https://github.com/microsoft/LightGBM/pull/3403) and possibly [on Linux in the future](https://github.com/microsoft/LightGBM/issues/4684).

> As for Intel GPU, I am not sure, are there any powerful GPUs that are widely used?

Probably, they **will**:

> Currently we know Intel plans to launch the first Arc GPUs for desktops as soon as the summer of 2022, and at the end of its promo video for the first Arc 3 laptop GPUs the company included a teaser image (above) of a full-size Arc desktop GPU. It sure looks like to be as big as an [Nvidia GeForce RTX 3060](https://www.tomsguide.com/news/nvidia-geforce-rtx-3060) — will it perform as well too? We'll have to wait and see.
   https://www.tomsguide.com/news/intel-arc-gpu-specs-release-date-features-and-more

**Comment by guolinke** [comment]
@StrikerRUS the code conversion should be manually done by our, and commit to repo. 
refer to

**Comment by StrikerRUS** [comment]
@guolinke 
Ah, I see now. That changes a lot!

**Comment by jameslamb** [comment]
Since the previous release, `{lightgbm}` is now depended on by several more packages.

<img width="409" alt="image" src="https://user-images.githubusercontent.com/7608904/179240445-8bf50847-30fe-4805-90cf-d67a9db28813.png">

https://cran.r-project.org/web/packages/lightgbm/index.html

Including a fairly high-profile one, `{bonsai}`, from RStudio: https://www.tidyverse.org/blog/2022/06/bonsai-0-1-0/.

I'm planning to try to do some work on those projects to make them compatible with v3.3.2 AND v4.0.0.

**Comment by jameslamb** [comment]
I've opened issues with offers of support for all of `{lightgbm}`'s reverse dependencies.

* `{cbl}`: https://github.com/dswatson/cbl/issues/1
* `{bonsai}`: https://github.com/tidymodels/bonsai/issues/34
* `{cheem}`: https://github.com/nspyrison/cheem/issues/2
* `{EIX}`: https://github.com/ModelOriented/EIX/issues/7
* `{fastshap}`: https://github.com/bgreenwell/fastshap/issues/47
* `{SHAPforxgboost}`: https://github.com/liuyanguu/SHAPforxgboost/issues/33
* `{shapviz}`: https://github.com/mayer79/shapviz/issues/22

I also want to document this small function in R (proposed in https://github.com/tidymodels/bonsai/pull/42) that can be used to check for the installed version of `{lightgbm}`:

```r
using_newer_lightgbm_version <- function(){
    utils::packageVersion("lightgbm") > package_version("3.3.2")
}

# example
if (using_newer_lightgbm_version()) {
    predict(bst, X, rawscore = TRUE)
} else {
    predict(bst, X, type = "raw")
}
```

------

I hope we can get v4.0.0 out soon, and that we'll start releasing more frequently in the future so that we can just use a cycle of deprecation warnings to communicate such changes, instead of needing to go submit patches.

**Comment by jameslamb** [comment]
@shiyu1994 @guolinke can you please devote some time to LightGBM over the next few weeks, and can we try to get a 4.0 release out?

At this point, it's been more than a year since the last release with substantive changes (v3.3.1, in October 2021). We really need your help.

**Comment by shiyu1994** [comment]
@jameslamb  And thanks for your support all the time.

**Comment by jameslamb** [comment]
Of course! At this point, since it's been more than a year since the last release, I think we need to accept that all the things we'd hoped to get into v4.0.0 aren't going to make it there. It would be better to get a v4.0.0 out and then plan other breaking changes for a v5.0.0 in the future, in my opinion, then to delay v4.0.0 another 6+ months.

**Comment by shiyu1994** [comment]
> It would be better to get a v4.0.0 out and then plan other breaking changes for a v5.0.0 in the future, in my opinion, then to delay v4.0.0 another 6+ months.

Strongly agree with that.

**Comment by jameslamb** [comment]
I'm focusing most of my attention right now on fixing the issues related to Python packaging and CI as soon as possible, in preparation for this release.

It has gotten really complicated, so typing out here the sequence that things need to happen.

1. get CI working again:
    - [ ] https://github.com/microsoft/LightGBM/pull/5588
2. merge integrated aarch64 wheels PR:
    - [ ] https://github.com/microsoft/LightGBM/pull/5252
3. switch to `manylinux_2_28` images for building Linux wheels
    - [ ] https://github.com/microsoft/LightGBM/pull/5580
    - [ ] https://github.com/guolinke/lightgbm-ci-docker/pull/27
4. merge all the new `guolinke/lightgbm-ci-docker` stuff to `master` in that repo
    - [ ] https://github.com/guolinke/lightgbm-ci-docker/pull/28
    - [ ] another PR to replace `*-dev` image tags in `.vsts-ci.yml` here in LightGBM
5. [ ] update to `macOS-11` images on Azure DevOps and build macOS wheels with newer XCode
    - ⚠️ **THIS HAS TO BE DONE BY DECEMBER 1, 2022** ⚠️ 
    - [ ] https://github.com/microsoft/LightGBM/pull/5396
6. switch from `setup.py` to `pyproject.toml` for wheel building or put a version ceiling on `pip` in CI
    - [ ] https://github.com/microsoft/LightGBM/issues/5061#issuecomment-1316317859

I'm happy to try to do most of the work, but really hope @StrikerRUS will be available to provide his expertise.

If he isn't, then @jmoralez @shiyu1994 @guolinke I will really need your help with reviews on these. If we are really struggling I can also try to recruit a volunteer from outside of LightGBM who's very familiar with Python packaging.

**Comment by guolinke** [comment]
@jameslamb Thank you! ping me when you need reviews.

**Comment by shiyu1994** [comment]
Just added quantized training (#5606) in the TODO list.

**Comment by jameslamb** [comment]
> We will make cuda_exp version supporting as many platforms as now by LightGBM. However, before cuda_exp is stable enough, I think we can still maintain the current CUDA and GPU versions, including bug fixing

I'm happy to see the progress on implementing more objective functions in the `cuda_exp` version.

@shiyu1994 @guolinke Do you think we should just remove the version that's currently called `cuda` and just focus on `cuda_exp`? The in LightGBM v4.0.0, users who are installing with `device_type = "cuda"` will get what we're now calling `"cuda_exp"`.

I think we should do that at this point, to:

* reduce maintenance burden in the repo
* ensure that we get bug reports only for the new CUDA version
* reduce the time it takes CI to run

I don't think this project is healthy enough right now (e.g. releasing frequently enough) to do something like release v4.0.0 with both `cuda` and `cuda_exp`, then support both for a long time, then eventually remove the old `cuda` code.

**Comment by shiyu1994** [comment]
> Do you think we should just remove the version that's currently called cuda and just focus on cuda_exp? The in LightGBM v4.0.0, users who are installing with device_type = "cuda" will get what we're now calling "cuda_exp".

Yes. We can do that in v4.0.0.

**Comment by jameslamb** [comment]
Excellent, thanks! I just opened #5677 for your consideration.

**Comment by jameslamb** [comment]
The last thing I absolutely want to get done for v4.0.0 is fix #5061, updating the way `lightgbm`'s wheels and sdists are built.

I'm working on that in #5759, and have started opening up some other small PRs to help with it (#5761, #5762).

@shiyu1994 @guolinke what else do you think *absolutely needs to be done* before we can put up a v4.0 release?

**Comment by shiyu1994** [comment]
@jameslamb For objective functions and metrics on new CUDA version, we will make most of them available on GPU. We may leave some of them (which are not very frequently used) falling back to CPU. This part we've almost done.

In addition, I think we should include quantized training. The code is ready in https://github.com/Quantized-GBDT/Quantized-GBDT. We just need to merge it into the official repo. (1~2 weeks ETA)

Finally, if possible, we may have a initial support for multi-GPU training on CUDA. An initial support of multi-GPU training may again takes 1~2 weeks. Depending on the progress of development and the date we want to make 4.0.0 released, we may adjust this goal.

**Comment by jameslamb** [comment]
Ok works for me! I'm happy to help with reviews, docs, tests, whatever you need.

**Comment by jameslamb** [comment]
There are now 1.5+ years worth of fixes and improvements sitting on `master` which haven't yet made it into a release.

@shiyu1994 @guolinke @jmoralez I think now that #5061 has been addressed and CPU-based quantized training has been added (#5800) we should finally do a v4.0.0 release and not wait any longer.

Do you agree? If yes, I'll create the release PR and we can talk more there.

**Comment by guolinke** [comment]
@shiyu1994 Do we have any other breaking changes?

**Comment by shiyu1994** [comment]
@jameslamb @guolinke What about we at least have #5933 merged into v4.0.0?

**Comment by jameslamb** [comment]
That PR doesn't contain breaking changes, so I don't think we should wait for it. It could be part of a 4.1 or 4.0.1 release.

If you can't think of any other *breaking* changes, I think we should move forward with v4.0.0.

**Comment by jameslamb** [comment]
I have one more small breaking change that is ready to be reviewed:

#5947 

After that, I'll make a PR for the 4.0 release. Exciting!!!

**Comment by jameslamb** [comment]
For anyone subscribed to notifications and wondering why v4.0.0 of the R package isn't on CRAN yet... it won't be until at least mid-August.

See https://github.com/microsoft/LightGBM/pull/5952#issuecomment-1646174008
